src/Raven.Server/Documents/Sharding/Queries/AbstractShardedQueryProcessor.cs

```diff
@@ -252,13 +314,48 @@ private BlittableJsonReaderObject RewriteQueryIfNeeded(BlittableJsonReaderObject
             }
         }
 
-        if (rewriteForPaging == false && rewriteForProjectionFromMapReduceIndex == false && rewriteForProjectionFromAutoMapReduceIndex == false)
+        if (rewriteForPaging == false 
+            && rewriteForProjectionFromMapReduceIndex == false 
+            && rewriteForProjectionFromAutoMapReduceQuery == false
+            && rewriteForFilterForMapReduceIndexOrAutoMapReduceQuery == false
+            && rewriteForLimitWithOrderByInMapReduce == false
+            && setOrderByFieldsForMapReduce == null)
             return queryTemplate;
 
         var clone = Query.Metadata.Query.ShallowCopy();
 
         DynamicJsonValue modifications = new(queryTemplate);
 
+        if (rewriteForFilterForMapReduceIndexOrAutoMapReduceQuery)
+        {
+            clone.Filter = null;
+            clone.FilterLimit = null;
+        }
+
+        if (rewriteForLimitWithOrderByInMapReduce)
+        {
+            clone.Offset = null;
+            clone.Limit = null;
+            clone.OrderBy = null;
+
+            modifications.Remove(nameof(IndexQueryServerSide.Start));
+            modifications.Remove(nameof(IndexQueryServerSide.PageSize));
+
+            rewriteForPaging = false;
+        }
+        else if (setOrderByFieldsForMapReduce != null)
+        {
+            clone.OrderBy = new List<(QueryExpression Expression, OrderByFieldType FieldType, bool Ascending)>();
+            Query.Metadata.OrderBy = new OrderByField[setOrderByFieldsForMapReduce.Count];
```

Comment:
isn't this done on a cached instance that can be shared between queries?

Reply:
done